Fix NH-3150: select id generator improvements

fredericDelaporte opened the pull request:
Fixes #822 
Replaces #110 

The select id generator generates the id by getting it from the database with a select query filtered by an unique key. This unique key could only be a single entity property, forcing to use a component if the key has to spawn many properties. The first commit add support for keys spanning many properties without using a component.

The select id generator was not supporting idbags, although some preliminary work was already done for this. The second commit fixes this.

These two commits are based on #110 commit, rebased then split and altered on many points. (Breaking changes removed, redundant casts removed, syntax modernized, multiple properties support added to the key parameter and not only to natural-id, tests adjusted, ...)

hazzik commented on src/NHibernate/Id/SelectGenerator.cs:
```diff
@@ -72,19 +81,42 @@ public partial class SelectGeneratorDelegate : AbstractSelectingDelegate
+
+					uniqueKeyTypes = uniqueKeyPropertyNames.Select(p => entityPersister.GetPropertyType(p)).ToArray();
+				}
+				else if (persister is AbstractCollectionPersister collectionPersister)
```
Probably should be `IQueryableCollection` instead

fredericDelaporte replied:
Indeed, amended.

fredericDelaporte commented:
Rebased. (The async generator update does not change anything to this PR added tests.)

hazzik commented on src/NHibernate/Id/SelectGenerator.cs:
```diff
@@ -72,19 +81,42 @@ public partial class SelectGeneratorDelegate : AbstractSelectingDelegate
+						$"Persister type {persister.GetType()} is not supported by Select Generator");
+				}
+
+				idSelectString = persister.GetSelectByUniqueKeyString(uniqueKeyPropertyNames);
```
After a closer look, I think this all shall be hidden inside `GetSelectByUniqueKeyString`

hazzik commented on src/NHibernate/Id/SelectGenerator.cs:
```diff
@@ -72,19 +81,42 @@ public partial class SelectGeneratorDelegate : AbstractSelectingDelegate
+				else if (persister is IQueryableCollection collectionPersister)
+				{
+					uniqueKeyPropertyNames = new string[2];
+					uniqueKeyTypes = new IType[2];
```
Are the `uniqueKeyPropertyNames` and `uniqueKeyTypes` set in this path used at all?

hazzik commented:
Overall solution looks like a hack (I'm not talking here about the workarounds to avoid breaking changes). It seems there could be a more elegant solution.

fredericDelaporte commented on src/NHibernate/Id/IPostInsertIdentityPersister.cs:
```diff
@@ -42,4 +68,23 @@ public interface IPostInsertIdentityPersister
+	/// An <see cref="IPostInsertIdentityPersister" /> that supports selecting by an unique key spanning
+	/// multiple properties.
+	/// </summary>
+	public interface IMultiPropertiesPostInsertIdentityPersister
```
This is not the identity which is composite, but the key used to retrieve the id. So maybe `ICompositeKeyPostInsertIdentityPersister`.

fredericDelaporte commented:
That is a solution attempting minimal impact, likely because the author has not contributed much to NHibernate. I could have reworked it a bit more.

fredericDelaporte commented:
Rebased, and fixup commits addressing review added.

fredericDelaporte commented on src/NHibernate/Id/SelectGenerator.cs:
```diff
@@ -95,31 +71,65 @@ protected internal override SqlString SelectSQL
+				if (persister is IEntityPersister)
+				{
+#pragma warning disable 618
+					BindParameters(session, ps, binder.Entity);
```
This is required to avoid a possible breaking change in the unlikely case someone uses a class overriding the obsoleted method. This causes `BindSelectByUniqueKey` of entity persisters to be currently not used.

Still more unlikely, this is maybe not enough: what if someone had implemented deriving classes correctly supporting idbag collections? But I do not see how we could handle that.

fredericDelaporte commented:
Rebased again. Since it is also about unique key, we could fear some conflicts with changes done for associations with property-ref. But there are none, because here this is about natural-ids, not property-ref.

As proposed on the [development group](https://groups.google.com/forum/#!topic/nhibernate-development/Y6ZvtgnoAns), and since no continued grounded disagreement has been expressed there since one week after raised concerns being addressed, I intend to merge this PR next week (likely on Wednesday).

(I am issuing this notice on PR issued before starting to apply this handling of merges. I will not do it for new PR.)

hazzik commented:
The issue noted that this solution is a hack has not been resolved since I posted it.

fredericDelaporte commented:
> The issue noted that this solution is a hack has not been resolved since I posted it.

In June:

> Rebased, and fixup commits addressing review added.

I had reworked it right after your post. But now the fixups are squashed.

hazzik commented on src/NHibernate/Id/IPostInsertIdentityPersister.cs:
```diff
@@ -31,6 +104,8 @@ public interface IPostInsertIdentityPersister
 		/// </param>
 		/// <returns> The SQL select string </returns>
+		// Since 5.2
+		[Obsolete("Have the persister implement IMultiPropertiesPostInsertIdentityPersister and use its GetSelectByUniqueKeyString(string[] suppliedPropertyNames, out IType[] parameterTypes).")]
```
The comment has to change to a new name

fredericDelaporte replied:
Done.

(And rebased, commits directly amended.)
